ambari-server/src/main/java/org/apache/ambari/server/serveraction/upgrades/DeleteUnsupportedServicesAndComponents.java

```diff
@@ -0,0 +1,92 @@
+import org.apache.ambari.server.state.ServiceComponentSupport;
+import org.apache.ambari.server.state.UpgradeContext;
+import org.apache.commons.lang.StringUtils;
+
+import com.google.inject.Inject;
+
+/**
+ * Upgrade Server Action that deletes the components and services which are no longer supported in the target stack.
+ * The deletable component or service should be in deletable state (stopped) before executing this.
+ */
+public class DeleteUnsupportedServicesAndComponents extends AbstractUpgradeServerAction {
+  @Inject
+  private ServiceComponentSupport serviceComponentSupport;
+
+  @Override
+  public CommandReport execute(ConcurrentMap<String, Object> requestSharedDataContext) throws AmbariException, InterruptedException {
+    Cluster cluster = getClusters().getCluster(getExecutionCommand().getClusterName());
+    UpgradeContext upgradeContext = getUpgradeContext(cluster);
+    Set<String> removedComponents = deleteUnsupportedComponents(cluster, upgradeContext.getRepositoryVersion());
+    Set<String> removedServices = deleteUnsupportedServices(cluster, upgradeContext.getRepositoryVersion());
+    return createCommandReport(0, HostRoleStatus.COMPLETED, "{}",
+      "Removed services: " + StringUtils.join(union(removedComponents, removedServices), ", "), "");
+  }
+
+  private Set<String> deleteUnsupportedServices(Cluster cluster, RepositoryVersionEntity repoVersion) throws AmbariException {
+    Set<String> servicesToBeRemoved = serviceComponentSupport.unsupportedServices(cluster, repoVersion.getStackName(), repoVersion.getStackVersion());
+    for (String serviceName : servicesToBeRemoved) {
+      cluster.deleteService(serviceName, new DeleteHostComponentStatusMetaData());
+      deleteUpgradeHistory(cluster, history -> serviceName.equals(history.getServiceName()));
+    }
+    return servicesToBeRemoved;
+  }
+
+  private Set<String> deleteUnsupportedComponents(Cluster cluster, RepositoryVersionEntity repoVersion) throws AmbariException {
+    Set<String> deletedComponents = new HashSet<>();
+    for (ServiceComponent component : serviceComponentSupport.unsupportedComponents(cluster, repoVersion.getStackName(), repoVersion.getStackVersion())) {
+      cluster.getService(component.getServiceName()).deleteServiceComponent(component.getName(), new DeleteHostComponentStatusMetaData());
+      deleteUpgradeHistory(cluster, history -> component.getName().equals(history.getComponentName()));
+      deletedComponents.add(component.getName());
+    }
+    return deletedComponents;
+  }
+
+  private void deleteUpgradeHistory(Cluster cluster, Predicate<UpgradeHistoryEntity> predicate) {
+    UpgradeEntity upgradeInProgress = cluster.getUpgradeInProgress();
+    List<UpgradeHistoryEntity> removed = upgradeInProgress.getHistory().stream()
+      .filter(each -> each != null && predicate.test(each))
+      .collect(toList());
+    removed.forEach(each -> upgradeInProgress.removeHistory(each));
```

Comment:
could just do foreach at stream instead of collectiong a list

Reply:
Actually I put it back because doing it in one step will throw ConcurrentModificationException.